### PR TITLE
[KYUUBI #3385] Set executor pod name prefix if missing in spark on k8s case

### PR DESCRIPTION
**Pull request by zwangsheng**
### _Why are the changes needed?_

See more in #3590
For #3590 this PR is reverted, and author not reply for long time.
Fix this issue #3385 here.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 1]
```diff
@@ -71,6 +71,12 @@
             <scope>provided</scope>
         </dependency>
 
+        <dependency>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/pom.xml:74]: Hmm?for what?

Comment (zwangsheng) [reply externals/kyuubi-spark-sql-engine/pom.xml:74]: Still in progress 😢

**Comment by ulysses-you** [comment]
shall we do this in `EngineRef` ? I remember `SparkProcessBuilder` also do validation for pod name.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineSuite.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.spark
+
+import org.apache.kyuubi.KyuubiFunSuite
+
+class SparkSQLEngineSuite extends KyuubiFunSuite {
+  private val EXECUTOR_POD_NAME_PREFIX_MAX_LENGTH = 237
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineSuite.scala:23]: why define again?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 3]
```diff
@@ -342,4 +351,23 @@ object SparkSQLEngine extends Logging {
...
+
+  @VisibleForTesting
+  def generateExecutorPodNamePrefixForK8s(userName: String): String = {
+    val resolvedUserName =
+      userName.trim.toLowerCase(Locale.ROOT)
+        .replaceAll("[^a-z0-9\\-]", "-")
+        .replaceAll("-+", "-")
+        .replaceAll("^-", "")
+    val podNamePrefixWithUser = s"kyuubi-$resolvedUserName-${Instant.now().toEpochMilli}"
+    if (podNamePrefixWithUser.length >= EXECUTOR_POD_NAME_PREFIX_MAX_LENGTH) {
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:363]: = is safe

**Review by pan3793 — APPROVED** [review]
lgtm, only minor comments

**Comment by zwangsheng** [comment]
> shall we do this in `EngineRef` ? I remember `SparkProcessBuilder` also do validation for pod name.

For validation, it looks more like to help fail fast on the Kyuubi side and is not supposed to help configure spark conf.

We do a lot of operations on `SparkSQLEngine` to help users optimize their configuration, so it makes sense to put this here.

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**